Add ArrayList

wgj opened the pull request:
`io.wgj.ctci.ArrayList` is modeled after [`java.util.ArrayList`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ArrayList.html), but only using methods that I thought were necessary to replace the internal `Object[]` arrays in `io.wgj.ctci.Queue` and `io.wgj.ctci.Stack`.

wgj committed:
Add ArrayList

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+    var newCapacity = capacity*2;
+    var newElements = new Object[newCapacity];
+    var i = 0;
+    for (Object e : elements) {
```
While this is correct, it would be customary to use the builtin arraycopy method here because it is very fast.

https://www.geeksforgeeks.org/system-arraycopy-in-java/

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+  transient Object[] elements;
+  static final int STARTING_CAPACITY = 1;
+  private int size;
+  private int capacity;
```
I believe "capacity" will always be the same as "elements.length", and therefore redundant to track separately.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.Iterator;
+
+public class ArrayList <E> implements Iterable<E> {
+  transient Object[] elements;
```
Did you intended to make this class serializable?  Regardless I don't think the transient keyword belongs on this.

wgj replied:
No. Thoughtlessly stolen from [here](https://www.baeldung.com/java-generic-array#1-using-arraylist-in-place-of-an-array). In hindsight, it seems like an language optimization that doesn't apply.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  private void grow() {
+    var newCapacity = capacity*2;
```
Nit: add spaces around boolean operators

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+   private void rangeCheck(int index) {
+    if (index >= size) {
```
This methods doesn't check for an index < 0, though maybe you intended to rely on the underlying array to throw in that case?  Inquiring minds must know.

wgj replied:
I stole it from [here](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/util/ArrayList.java#l636), and not until you asked this question did I read the very next method.

Fwiw, I looked up that file to see how they were dealing with type erasure and `Object[]`s, and stopped reading after I realized how easy it was accidentally copy from the implementation.

At the expense of tainting learning by seeing how it's really done, it was comforting to see that they to had to deal with unchecked warnings just like me, and there wasn't some higher level Java that only a JDK writer gets to use.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+      public boolean hasNext() {
+        try {
+          rangeCheck(index+1);
+        } catch (IndexOutOfBoundsException e) {
```
I normally wouldn't recommend code duplication, but exceptions are laboriously slow and arrays need to be fast so I'd recommend checking the index instead of catching the exception.

sabeechen replied:
I'd still recommend this but its more of a nit than anything else.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+    // elements after the removed index shift to the left.
+    for (int i = index; i < size-1; i++) {
+      elements[i] = elements[i+1];
+    }
```
Recommend also setting "elements[size] = null" so that java's pathetic reference counting excuse for a garbage collector can clean up the object.  Otherwise the runtime would think that removed objects are still needed until the arraylist grows to overtake them again.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+    if (size != 0) {
+      return false;
+    }
+    return true;
```
return size == 0

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  public boolean add(E e) {
+    if (size >= capacity) {
```
Please justify to me that this should be >= rather than ==.  I'm not saying that this is wrong, I just want you to convince me.

wgj replied:
I think it's no longer relevant. When I wrote written, I must've thought `size` might be incremented elsewhere, and I was worried I'd confuze number of elements and number of indicies preventing `grow()` from happening. I reasoned that a slightly premature `grow()` was better than a failure to call `grow()`.

sabeechen replied:
What I wanted to get at is that the way this class is written it is impossible for ```size``` > ```capacity```, so in some sense you're adding unreachable/untestable code by using ```>=``` instead of just ```==```.

That being said, I see a lot of people still do this in this case because they feel its programming defensively, I guess against cosmic rays flipping RAM bits or something.  Personally when I see something like this I assume I misunderstand the class and go re-read everything, wasting time.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,105 @@
+      @Override
+      public boolean hasNext() {
+        try {
+          rangeCheck(index+1);
```
I believe this should be `rangeCheck(index)`, since `next()` returns `elementData(index)`.  The tests miss this.

sabeechen replied:
I think this is still a problem, because the test for it isn't checking the length of the returned items from the iterator.

wgj committed:
Refactor to use io.wgj.ctci.ArrayList

Rename methods to queue specific names

Remove unused or unneeded methods and fields.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -0,0 +1,136 @@
+
+  @Test
+  public void arrayList_new() {
+    ArrayList<Integer> al = new ArrayList<>();
```
Your desire to _truly_ "unit test" by having a test for each method is admirable but imho prioritizing the intention over practicality.  For example every method in this test calls the initializer, so removing this test wouldn't lose coverage and also wouldn't make it harder to track down problems if some issue came up with the initializer, so why keep it around?

IMHO, in practice unit tests are best when they cover a "unit of functionality" rather than any kind of more hard-lined division.  For example I think it would be reasonable to have a unit test that just verifies initialization, add, get, and size all together.  In that case you'd be writing a test for `add()` that verifies it worked by checking the output from `size()` and `get()`.

wgj replied:
Historically, it's a bad habit, but in this case, it's here from casual TDD; I was making successively larger tests as the code under test was being written. I need to do a better job of cleaning up those _now_ useless tests.

Also, as a preference, I find JUnit's output difficult to intuit which assertion failed when there's more than a few in a single test. I believe I'm getting better at that, and need this training wheel less.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -0,0 +1,136 @@
+    assertEquals(0, al.size());
+    al.add(1);
+    assertEquals(1, al.size());
+  }
```
This is a perfect example of what I'm talking about.  If you have two tests that that are exactly the same except one has extra statements, it almost always the case that one of the them is unnecessary.  Coverage can actually be just duplication if it doesn't help you find problems.

wgj replied:
Gotcha. I'm starting to think of tests more for their coverage than as discrete little experiments.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -0,0 +1,136 @@
+
+  @Test
+  public void arrayList_grow() {
+    ArrayList<Integer> al = new ArrayList<>();
```
Nit: not everyone would agree with this but for the sake of communicating intent in code I'd avoid abbreviations when space permits it because they are often ambiguous.  In this case I think "list" would be a good replacement.

wgj replied:
Thanks! `list` is a good compromise and falls on the right side of terse.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -0,0 +1,136 @@
+    assertEquals(add_ints[0], al.remove(0));
+
+    Integer[] get_ints = {2, 3, 4, 5};
+    for (int i = 0; i <= get_ints.length - 1; i++) {
```
I think in this comparison could just be i < get_ints.length?

Also consider making a helper method that takes an ArrayList and an array, ensuring the contents are equivalent.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -0,0 +1,136 @@
+    assertEquals((Integer) 0, al.get(0));
+    al.set(0, 42);
+    assertEquals((Integer) 42, al.get(0));
+  }
```
Nit: Use consistent whitespace between test methods.

wgj committed:
Refactor Stack to use io.wgj.ctci.ArrayList

Remove unused or unneeded methods and fields.

wgj committed:
Whitespace

wgj committed:
Whitespace

wgj committed:
Reduce tests and merge coverage

wgj committed:
Reduce tests and merge coverage

wgj committed:
Feedback from code review

wgj committed:
Feedback from code review

wgj committed:
Replace al with list

wgj committed:
Replace al with list

wgj committed:
Remove unused import

wgj committed:
Remove unused import

wgj committed:
Fix comparison

wgj committed:
Fix comparison

wgj committed:
Fix assertions

wgj committed:
Fix assertions

wgj committed:
Remove inaccessible elements

wgj committed:
Remove inaccessible elements

wgj committed:
Use arraycopy()

wgj committed:
Use arraycopy()

wgj committed:
Remove unused modifier

wgj committed:
Remove unused modifier

wgj committed:
Merge branch 'refactor_arraylist' into add_arraylist

wgj commented:
@sabeechen added the commits with your feedback to this PR. Sorry, they were on the wrong branch. The scope of this PR is a little wider than I intended now, but it's all stuff I wanted you to look at anyways.

sabeechen commented on src/main/java/io/wgj/ctci/ArrayList.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Iterator;
+
+public class ArrayList <E> implements Iterable<E> {
+  Object[] elements;
```
I think this should be private

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-  int size;
+  ArrayList<E> list;
+  int frontIndex;
+  int backIndex;
```
Nit: java's default access modifier is package-private, which leaves the internals of this class visible to other classes in the package.  If that is your intention this is fine, but if not a good practice is to always make members private unless you have a need not to for better encapsulation.

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-  public E remove() {
-    if (size == 0) {
-      throw new IndexOutOfBoundsException();
+  public void dequeue() {
```
Its customary for pop/dequeue/remove/etc to return the element removed.  It usually makes using the object a lot cleaner.

sabeechen replied:
This method should indicate if an item was dequeued or not.  I think you could do this by either 

- returning true/false if an item is dequeued.
- throwing a NoSuchElementException if its empty (this is more common)

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-    if (size == 0) {
-      throw new IndexOutOfBoundsException();
+  public void dequeue() {
+    if (! isEmpty()) {
```
Nit: extra whitespace

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-      throw new IndexOutOfBoundsException();
+  public void dequeue() {
+    if (! isEmpty()) {
+      frontIndex++;
```
You'll probably want to set list[frontIndex] = null for the same garbage collection reasons.

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-  public E peek() {
-    if (size == 0) {
+  public E front() {
+    if (frontIndex == backIndex) {
```
The problem with returning null here is you wouldn't be able to tell the difference between an empty queue or just the head element being null.  I suppose you could check the size if you get null, but its kind of an awkward contract.  
Some argue that using null to represent any state other than "this value has not been assigned yet" leads to messy and ambiguous code.  I think they have a point.
In this instance I'd recommend throwing a NoSuchelementException

sabeechen replied:
This check could just be isEmpty()

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-    head = 0;
-    tail = 0;
-    size = 0;
+    list = new ArrayList<>();
```
Nit: you could just initialize these in the member definitions and then you wouldn't even need to define a constructor.

sabeechen commented on src/main/java/io/wgj/ctci/Queue.java:
```diff
@@ -2,88 +2,35 @@
-    size = 0;
+    list = new ArrayList<>();
+    frontIndex = 0;
+    backIndex = 0;
```
IIUC backIndex is always exactly list.size()?

sabeechen commented on src/main/java/io/wgj/ctci/Stack.java:
```diff
@@ -1,65 +1,31 @@
-        Object[] newStack = copyOf(stack, newCap);
-        capacity = newCap;
-        stack = newStack;
+        // This shouldn't been costly, since we're always removing from the top of the stack.
```
s/been/be

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -4,132 +4,108 @@
-    assertEquals(4, al.size());
+    assertEquals(ints[0], list.remove(0));
+    Integer[] ints_with_front_removed = {2, 3, 4, 5};
+    assertTrue(compareArrayToArrayList(ints_with_front_removed, list));
```
Recommend just passing the array directly, IE: compareArrayToArrayList(int[]{2, 3, 4, 5}, list));

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -4,132 +4,108 @@
 
+  private boolean compareArrayToArrayList(Integer[] array, ArrayList<Integer> list) {
+    if (array.length != list.size()) {
+      return false;
```
IMHO it would be cleaner to have this method do the asserts directly, ie assertEquals(array.length, list.size())
That would make tracking down failures easier too, since juint would show which line ti failed on.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -4,132 +4,108 @@
+      }
+    }
+    return true;
+  }
```
Nit: missing newline. Recommend using a linter.

sabeechen commented on src/test/java/io/wgj/ctci/ArrayListTest.java:
```diff
@@ -4,132 +4,108 @@
-    for (Integer i : al) {
-      assertEquals(al.get(index), i);
+    for (Integer i : list) {
+      assertEquals(list.get(index), i);
```
This also needs to verify that the length of items the iterator returns matches the expected length, I believe there is a bug here.
